WcaOnRails/app/models/team_member.rb

```diff
@@ -0,0 +1,20 @@
+class TeamMember < ActiveRecord::Base
+  belongs_to :team
+  belongs_to :user
+
+  attr_accessor :current_user
+
+  validate :cannot_demote_leader
+  def cannot_demote_leader
+    if end_date != nil && end_date < Date.today && team_leader
+      errors.add(:end_date, "A team leader must be a current member.")
+    end
+  end
+
+  validate :cannot_demote_oneself
+  def cannot_demote_oneself
+    if current_user == self.user_id && end_date != nil && end_date < Date.today
```

Comment:
That sounds good.

On Mon, Mar 21, 2016 at 4:48 AM, Jeremy Fleischman <notifications@github.com
> wrote:

> In WcaOnRails/app/models/team_member.rb
> <https://github.com/cubing/worldcubeassociation.org/pull/390#discussion_r56788818>
> :
>
> > +class TeamMember < ActiveRecord::Base
> > +  belongs_to :team
> > +  belongs_to :user
> > +
> > +  attr_accessor :current_user
> > +
> > +  validate :cannot_demote_leader
> > +  def cannot_demote_leader
> > +    if end_date != nil && end_date < Date.today && team_leader
> > +      errors.add(:end_date, "A team leader must be a current member.")
> > +    end
> > +  end
> > +
> > +  validate :cannot_demote_oneself
> > +  def cannot_demote_oneself
> > +    if current_user == self.user_id && end_date != nil && end_date < Date.today
>
> Very clean validations!
>
> I see you have this end_date != nil && end_date < Date.today in 2 places.
> How about pulling it out into a def current_member? or def ongoing_member?
> method? I feel like that would be easier to read.
>
> —
> You are receiving this because you authored the thread.
> Reply to this email directly or view it on GitHub
> <https://github.com/cubing/worldcubeassociation.org/pull/390/files/4cfc692de289d2d4d24e03be7e8db4b02da2a89d#r56788818>
>